src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets

```diff
@@ -68,13 +75,26 @@ Copyright (c) .NET Foundation. All rights reserved.
        ToolsSettingsFilePath="$(_ToolsSettingsFilePath)" />
   </Target>
 
-  <Target Name="ResolveApphostAsset" DependsOnTargets="ResolveReferences;Publish">
+  <!--
+    ============================================================
+                                        GenerateShimsAssets
+
+    Generate packaged shims for signing when PackAsToolShimRuntimeIdentifiers is set
+    ============================================================
+    -->
+
+  <Target Name="GenerateShimsAssets"
+          AfterTargets="ResolveReferences"
```

Comment:
`AfterTargets` is fine, with the caveat that you shouldn't do `AfterTargets="Build"` because of https://github.com/Microsoft/msbuild/issues/3345. But I generally prefer to use `BeforeTargets`--can you `BeforeTargets="IncrementalClean"`?

Reply:
@rainersigwald thank, although the task itself depends on `ResolveReferences`. Should I just dependson =`ResolveReferences`. and `BeforeTargets="IncrementalClean"`. But I think use dependson with before/after target is kind of strange.

Reply:
It's reasonable to have both `BeforeTargets="IncrementalClean" DependsOnTargets="ResolveReferences"`. I'd go with that.

It is funny looking, but they're different relationships: BeforeTargets means "before you run that, run me first", while DependsOn means "before I run, run these first". Both can be true at once.

Reply:
I would expect the logic for this target to be the same as GenerateBuildDependencyFile. We went through a few iterations on that and ended up against beforetargets=incrementalclean IIRC. It would be good to have one pattern for extra files we create.

Reply:
I see, I read the comment. It makes sense. Now the problem is, how to set the timing(before or after what?) I think `BeforeTargets="CopyFilesToOutputDirectory"` is reasonable.

Changed to following, I will push later but I don't want to hide the thread.

```xml
  <Target Name="GenerateShimsAssets"
          BeforeTargets="CopyFilesToOutputDirectory"
          DependsOnTargets="ResolveReferences"
          Condition="'$(PackAsTool)' == 'true'" >

    <ItemGroup>
      <_PackAsToolShimRuntimeIdentifiers Include="$(PackAsToolShimRuntimeIdentifiers)"/>
    </ItemGroup>

    <PropertyGroup>
      <PackagedShimOutputRootDirectory Condition=" '$(PackagedShimOutputRootDirectory)' == '' ">$(OutDir)</PackagedShimOutputRootDirectory>
    </PropertyGroup>

    <GenerateShims
      DotNetAppHostExecutableNameWithoutExtension="$(_DotNetAppHostExecutableNameWithoutExtension)"
      PackagedShimOutputDirectory="$(PackagedShimOutputRootDirectory)/shims/$(TargetFramework)"
      PackageId="$(PackageId)"
      PackageVersion="$(Version)"
      ProjectAssetsFile="$(ProjectAssetsFile)"
      ProjectPath="$(MSBuildProjectFullPath)"
      ShimRuntimeIdentifiers="@(_PackAsToolShimRuntimeIdentifiers)"
      TargetFrameworkMoniker="$(NuGetTargetMoniker)"
      ToolCommandName="$(ToolCommandName)"
      ToolEntryPoint="$(ToolEntryPoint)"
      >

      <Output TaskParameter="EmbeddedApphostPaths" ItemName="_EmbeddedApphostPaths" />
    </GenerateShims>

    <ItemGroup>
      <!-- Do this in an ItemGroup instead of as an output parameter so that it still gets added to the item set
           during incremental builds when the task is skipped -->
      <FileWrites Include="@(_EmbeddedApphostPaths)"/>
    </ItemGroup>
  </Target>
```